homeassistant/components/thread/dataset_store.py

```diff
@@ -0,0 +1,150 @@
+    id: str = dataclasses.field(default_factory=uuid_util.random_uuid_hex)
+
+    @property
+    def extended_pan_id(self) -> str | None:
+        """Return extended PAN ID as a hex string."""
+        dataset = tlv_parser.parse_tlv(self.tlv)
+        return dataset.get(tlv_parser.MeshcopTLVType.EXTPANID)
+
+    @property
+    def network_name(self) -> str | None:
+        """Return network name as a string."""
+        dataset = tlv_parser.parse_tlv(self.tlv)
+        return dataset.get(tlv_parser.MeshcopTLVType.NETWORKNAME)
+
+    @property
+    def pan_id(self) -> str | None:
+        """Return PAN ID as a hex string."""
+        dataset = tlv_parser.parse_tlv(self.tlv)
+        return dataset.get(tlv_parser.MeshcopTLVType.PANID)
+
+    def to_json(self) -> dict[str, Any]:
+        """Return a JSON serializable representation for storage."""
+        return {
+            "created": self.created.isoformat(),
+            "id": self.id,
+            "preferred": self.preferred,
+            "source": self.source,
+            "tlv": self.tlv,
+        }
+
+
+class DatasetStore:
+    """Class to hold a collection of thread datasets."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the dataset store."""
+        self.hass = hass
+        self.datasets: dict[str, DatasetEntry] = {}
+        self._store: Store[dict[str, list[dict[str, Any]]]] = Store(
+            hass,
+            STORAGE_VERSION_MAJOR,
+            STORAGE_KEY,
+            atomic_writes=True,
+            minor_version=STORAGE_VERSION_MINOR,
+        )
+
+    @callback
+    def async_add(self, source: str, tlv: str) -> None:
+        """Add dataset, does nothing if it already exists."""
+        # Make sure the tlv is valid
+        tlv_parser.parse_tlv(tlv)
+        # Bail out if the dataset already exists
+        if any(dataset for dataset in self.datasets.values() if dataset.tlv == tlv):
+            return
+
+        # Set to preferred if this is the first dataset
+        preferred = not bool(self.datasets)
+        dataset = DatasetEntry(preferred=preferred, source=source, tlv=tlv)
+        self.datasets[dataset.id] = dataset
+        self.async_schedule_save()
+
+    @callback
+    def async_delete(self, dataset_id: str) -> None:
+        """Delete dataset."""
+        if not (dataset := self.datasets.get(dataset_id)):
+            return
+        if dataset.preferred:
+            raise HomeAssistantError("attempt to remove preferred dataset")
+        del self.datasets[dataset_id]
+        self.async_schedule_save()
+
+    @callback
+    def async_get(self, dataset_id: str) -> DatasetEntry | None:
+        """Get dataset by id."""
+        return self.datasets.get(dataset_id)
+
+    async def async_load(self) -> None:
+        """Load the datasets."""
+        data = await self._store.async_load()
+
+        datasets: dict[str, DatasetEntry] = {}
+
+        if isinstance(data, dict):
```

Comment:
this is a bit weird check. We always put a `dict` in, so we should just check if it's not None.
```suggestion
        if data is not None:
```